lib/std/SinglyLinkedList.zig

```diff
@@ -164,3 +164,130 @@ test "basics" {
     try testing.expect(@as(*L, @fieldParentPtr("node", list.first.?.next.?)).data == 2);
     try testing.expect(list.first.?.next.?.next == null);
 }
+
+/// implements a "simple" intrusive singly linked list with a "data" field alongside
+/// "node" field.  This hides @fieldParentPtr complexity and adds type safety for simple
+/// cases.
+///
+/// note that the signatures on the member functions of the generated datastructure take
+/// pointers to the payload, not the node.
+pub fn Simple(T: type) type {
+    return struct {
+        const SimpleLinkedList = @This();
+        wrapped: SinglyLinkedList = .{},
+
+        pub const Payload = struct {
+            data: T,
+            node: Node = .{},
+
+            pub fn next(payload: *Payload) ?*Payload {
+                return @fieldParentPtr("node", payload.node.next orelse return null);
+            }
+
+            pub fn insertAfter(payload: *Payload, new_payload: *Payload) void {
+                payload.node.insertAfter(&new_payload.node);
+            }
+        };
+
+        pub fn prepend(list: *SimpleLinkedList, new_payload: *Payload) void {
+            list.wrapped.prepend(&new_payload.node);
+        }
+
+        pub fn remove(list: *SimpleLinkedList, payload: *Payload) void {
+            list.wrapped.remove(&payload.node);
+        }
+
+        /// Remove and return the first node in the list.
+        pub fn popFirst(list: *SimpleLinkedList) ?*Payload {
+            const poppednode = (list.wrapped.popFirst()) orelse return null;
+            return @fieldParentPtr("node", poppednode);
+        }
+
+        /// Given a Simple list, returns the payload at position <index>.
+        /// If the list does not have that many elements, returns `null`.
+        ///
+        /// This is a linear search through the list, consider avoiding this
+        /// operation, except for index == 0
+        pub fn at(list: *SimpleLinkedList, index: usize) ?*Payload {
+            var thisnode = list.wrapped.first orelse return null;
+            var ctr: usize = index;
+            while (ctr > 0) : (ctr -= 1) {
+                thisnode = thisnode.next orelse return null;
+            }
+            return @fieldParentPtr("node", thisnode);
+        }
+
+        // Iterate over all nodes, returning the count.
+        ///
+        /// This operation is O(N). Consider tracking the length separately rather than
+        /// computing it.
+        pub fn len(list: SimpleLinkedList) usize {
+            return list.wrapped.len();
+        }
```

Comment:
Would it not be better to just not have `.len()` to discourage use of such a fuction?

When using a linked-list this is not an operation one should really ever use (same goes for the intrusive linked-list) as it touches all nodes thus involves several cache-line misses in the worst case. It may be `O(N)` but the cost is not comparable to say the sum of an array which is also `O(N)` but likely involves fewer cache misses.

I'd say the same for `at` as well. Those should be rare or another structure should be chosen in most cases.

Reply:
well, if that's the case it then shouldn't be in base linked list implementation, either.  I think it's fine.  For debugging/asserts, if you know your lists are not that long, etc.